File: Java Blind 75/Arrays/Maximum Product Subarray.java

import java.util.Currency;
class Solution {
    public int maxProduct(int[] nums) {
        int maxSoFar = nums[0], minSoFar = nums[0], maxProd = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int tempMax = Math.max(nums[i], Math.max(maxSoFar * nums[i], minSoFar * nums[i]));
            minSoFar = Math.min(nums[i], Math.min(maxSoFar * nums[i], minSoFar * nums[i]));
            maxSoFar = tempMax;
            maxProd = Math.max(maxSoFar, maxProd);
        }
        return maxProd;
    }
}

File: Python Blind 75/Arrays/Maximum Product Subarray.py
class Solution(object):
    def maxProduct(self, nums):
        # Time: O(n), Space: O(1)
        """
        :type nums: List[int]
        :rtype: int
        """
        currMax, minMax, maxProd = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            tempMax = max(nums[i], max(nums[i] * currMax, nums[i] * minMax))
            minMax = min(nums[i], min(nums[i] * currMax, nums[i] * minMax))
            currMax = tempMax
            maxProd = max(maxProd, currMax)
        return maxProd
        